### PR TITLE
Guess s3 multipart upload content type

### PR DESCRIPTION
**Pull request by iurisilvio**
### Description

I added mimetype guessing to S3 multipart upload.

https://issues.apache.org/jira/browse/LIBCLOUD-958

### Status

Done, ready for merge.

### Checklist (tick everything that applies)

- [x] [Code linting](http://libcloud.readthedocs.org/en/latest/development.html#code-style-guide) (required, can be done after the PR checks)
- [ ] Documentation
- [x] [Tests](http://libcloud.readthedocs.org/en/latest/testing.html)

### CODE REVIEW

File: libcloud/test/storage/test_s3.py [hunk 1]
```diff
@@ -905,6 +907,29 @@ def test_upload_big_object_via_stream(self):
         self.assertEqual(obj.name, object_name)
         self.assertEqual(obj.size, CHUNK_SIZE * 3)
 
+    def test_upload_object_via_stream_guess_file_mime_type(self):
+        def dummy_content_type(name):
+            return 'application/zip', None
+
+        old_func = libcloud.utils.files.guess_file_mime_type
+        libcloud.utils.files.guess_file_mime_type = dummy_content_type
```
Comment (pquentin) [inline libcloud/test/storage/test_s3.py]: Can you please use monkeypatch.setattr from pytest? It's more robust: for example, if the test fails before entering the `try` block, we will have modified `libcloud.utils.files.guess_file_mime_type` for the other tests too.

File: libcloud/test/storage/test_s3.py [hunk 2]
```diff
@@ -905,6 +907,29 @@ def test_upload_big_object_via_stream(self):
...
+            self.mock_response_klass.type = None
+
+        container = Container(name='foo_bar_container', extra={},
+                              driver=self.driver)
+        object_name = 'foo_test_stream_data'
+        iterator = BytesIO(b('234'))
+        try:
+            self.driver.upload_object_via_stream(container=container,
+                                                 object_name=object_name,
+                                                 iterator=iterator)
```
Comment (pquentin) [inline libcloud/test/storage/test_s3.py]: non blocking nitpick: Is there a way to check that the correct content type was used during the upload?

**Review by pquentin — CHANGES_REQUESTED** [review]
Looks good, the only change to make is the `libcloud.utils.files.guess_file_mime_type` monkeypatch.

File: libcloud/test/storage/test_s3.py (hunk 2 above)
Comment (iurisilvio) [reply libcloud/test/storage/test_s3.py]: I tried, but `upload_object_via_stream` don't return anything useful to validate it. I'd have to mock `_initiate_multipart` to assert headers or extract all header generation to another method.

File: libcloud/test/storage/test_s3.py (hunk 1 above)
Comment (iurisilvio) [reply libcloud/test/storage/test_s3.py]: Changed to use Python `mock.patch`.

File: libcloud/test/storage/test_s3.py (hunk 2 above)
Comment (iurisilvio) [reply libcloud/test/storage/test_s3.py]: I changed this test to use a mock. I added a check to assert if my mock function is called. Not perfect, but at least we know the code is trying to guess the mimetype.

Comment (pquentin) [reply libcloud/test/storage/test_s3.py]: Great, thanks! Another non-blocking nitpick that really you don't have to follow: can we try to be more resilient to future typos here? See https://engineeringblog.yelp.com/2015/02/assert_called_once-threat-or-menace.html. I think using autospec would work in this case?

**Review by pquentin — APPROVED** [review]
Thanks! Will merge when I get the chance.

File: libcloud/test/storage/test_s3.py (hunk 2 above)
Comment (iurisilvio) [reply libcloud/test/storage/test_s3.py]: Great article. I had mock assert issues in the past, I know how it is painful to debug. Added `autospec=True` and it works fine. :+1:

**Comment by pquentin** [comment]
I appreciate that you took the time to improve your patch multiple times. Thanks! Merged in trunk. ✨

**Pull request closed without merging**